API: Remove the constant `JETPACK_CLIENT__HTTPS`.

sergeymitr opened the pull request:
Fixes #16980

#### Changes proposed in this Pull Request:
Constant `JETPACK_CLIENT__HTTPS` is used to explicitly switch outbound HTTPS requests to HTTP.
`public-api.wordpress.com` no longer supports HTTP requests, so Jetpack cannot function if outbound HTTPS requests do not work.

We introduce a new constant `JETPACK__WPCOM_JSON_API_BASE` which has the protocol (`https://`) hardcoded into it, and deprecates the existing `JETPACK__WPCOM_JSON_API_HOST` which requires the protocol to be attached to it separately.
We also clean up the code accordingly, and deprecate functions that are no longer needed.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
This is a janitorial PR, meaning we only need to confirm that everything works as intended.
1. Connect Jetpack, make sure Jetpack Dashboard and Settings load correctly.
2. Activate Jetpack, try to logout of WP and login back again.
3. Go to https://wordpress.com/themes/, choose your site, and install a Jetpack theme.
4. Open the theme's `style.css` and downgrade the theme's point version (e.g. `1.2.3` to `1.2.2`).
5. Go to your site's `/wp-admin/themes.php` and make sure that you're suggested to update the theme. Update it.
6. If everything of the above went smoothly, check the `/wp-content/debug.log` and make sure there are no deprecated errors: `grep deprecated wp-content/debug.log`

#### Proposed changelog entry for your changes:
* Cleaning up outdated code that allows non-secure requests to WP.com API, because WP.com no longer accepts such requests.

sergeymitr committed:
API: Remove the constant `JETPACK_CLIENT__HTTPS`.

Constant `JETPACK_CLIENT__HTTPS` is used to explicitly switch outbound HTTPS requests to HTTP.
`public-api.wordpress.com` no longer supports HTTP requests, so Jetpack cannot function if outbound HTTPS requests do not work.

We introduce a new constant `JETPACK__WPCOM_JSON_API_BASE` which has the protocol (`https://`) hardcoded into it, and deprecates the existing `JETPACK__WPCOM_JSON_API_HOST` which requires the protocol to be attached to it separately.
We also clean up the code accordingly, and deprecate functions that are no longer needed.

anomiex reviewed (commented):
I'm trying to start figuring things out by looking at PRs.

I see there's also still uses of the deprecated `JETPACK__WPCOM_JSON_API_HOST` constant in a few files
* [class.jetpack-cli.php](https://github.com/Automattic/jetpack/blob/d48ac90834169e7978d1838ec420b093f75fb27c/class.jetpack-cli.php#L17930)
* [functions.global.php](https://github.com/Automattic/jetpack/blob/d48ac90834169e7978d1838ec420b093f75fb27c/functions.global.php#L271)
* [modules/videopress/utility-functions.php](https://github.com/Automattic/jetpack/blob/d48ac90834169e7978d1838ec420b093f75fb27c/modules/videopress/utility-functions.php#L497)
* [_inc/class.jetpack-provision.php](https://github.com/Automattic/jetpack/blob/d48ac90834169e7978d1838ec420b093f75fb27c/_inc/class.jetpack-provision.php#L279)
* And a comment mentioning it in [modules/videopress/class.videopress-edit-attachment.php](https://github.com/Automattic/jetpack/blob/d48ac90834169e7978d1838ec420b093f75fb27c/modules/videopress/class.videopress-edit-attachment.php#L157)

kraftbj commented on modules/protect.php:
```diff
@@ -858,15 +856,15 @@ function get_transient( $transient ) {
+	 *
+	 * @deprecated 9.0.0 Use constant `JETPACK_PROTECT__API_HOST` instead.
+	 */
 	function get_api_host() {
```
I agree that if we're going to deprecate it, we should fully deprecate it -- remove all uses and use the `_deprecated_function()` call unless there is a particular reason for not doing so. For example, in the past, we delayed calling `_deprecated_function` when Akismet and VaultPress called a function that we were deprecating to give them time to update it. It was a known issue and they were the only known external users of that function, so there was no benefit to throw the deprecation notice into the PHP logs.

We tend to leave code in a bit longer—we're just now looking at removing things that have been deprecated for six months or more. With all of the potential ways that plugins/themes could be calling on some of the code (and depending on how long it has existed), we're a been conservative on actually removing deprecated code—especially if it basically just calls a new supported function.

sergeymitr replied:
@anomiex, @kraftbj 
You're right, I cut some corners to make the testing a little bit easier.
But since it's already unsafe to merge into `9.0`, I updated all those features and we'll merge it early in the next cycle to give more time for testing.

kraftbj commented on _inc/lib/debugger/class-jetpack-cxn-tests.php:
```diff
@@ -827,7 +827,7 @@ protected function last__wpcom_self_test() {
 
 		$self_xml_rpc_url = site_url( 'xmlrpc.php' );
 
-		$testsite_url = Connection_Utils::fix_url_for_bad_hosts( JETPACK__API_BASE . 'testsite/1/?url=' );
```
I am glad we can finally stop using this function. Thanks for cleaning this up.

sergeymitr committed:
API: Replacing deprecated `JETPACK__WPCOM_JSON_API_HOST` with JETPACK__WPCOM_JSON_API_BASE.

sergeymitr committed:
Bumping version to 9.1 for deprecated features.

We aren't getting it merged into 9.0, so I updated the code to deprecate the methods and constants since 9.1.

sergeymitr commented:
**The PR is aimed at Jetpack 9.1, please do not merge until the code freeze.**

kraftbj commented on packages/connection/src/class-client.php:
```diff
@@ -445,18 +441,13 @@ public static function _stringify_data( $data ) { // phpcs:ignore PSR2.Methods.M
-		 *
-		 * @param bool $proto Defaults to true.
-		 */
-		$https = apply_filters( 'jetpack_can_make_outbound_https', true );
```
Can you add this to the `$filter_deprecated_list` array in `class.jetpack.php`?

sergeymitr replied:
Done 👍

kraftbj commented on packages/connection/src/class-utils.php:
```diff
@@ -18,24 +18,16 @@ class Utils {
+	 *
+	 * @deprecated 9.1.0 Jetpack can't function properly on servers that don't support outbound HTTPS requests.
 	 */
 	public static function fix_url_for_bad_hosts( $url ) {
```
Minor but let's also remove the reference to this file `* @see Utils::fix_url_for_bad_hosts()` in the compat package.

sergeymitr replied:
Cleaned up 👍

kraftbj reviewed (changes requested):
A couple of minor nitpicks.

sergeymitr committed:
API: Cleaning up deprecated code.

- The filter 'jetpack_can_make_outbound_https' is no longer used. This commit adds the filter into the list of deprecated filters.
- Method `Automattic\Jetpack\Connection\Utils::fix_url_for_bad_hosts()` is now deprecated, so we're removing references to it in the PHPDoc blocks.
- Class `Automattic\Jetpack\Constants` is no longer used in the `Connection\Utils`, removing the `use` statement.
